### PR TITLE
ESM Support

### PR DESCRIPTION
**Pull request by wemeetagain**
Use ESM (only) across all lodestar packages.

- [x] Fix linter errors
- [x] Fix `check-types` errors
- [x] Fix merge sim tests
- [x] Fix sim tests
- [x] Fix e2e tests
- [x] Fix spec tests
- [x] Fix typescript-docs-verifier - disabled for now
- [x] Fix merge conflicts

### CODE REVIEW

File: packages/beacon-state-transition/.mocharc.yaml [hunk 1]
```diff
@@ -1,6 +1,8 @@
 colors: true
-require:
-  - ts-node/register
-  - ./test/setup.ts
```
Comment (dapplion) [inline packages/beacon-state-transition/.mocharc.yaml:4]: If the test setup is disabled, how is bls setup now? With the automatic top level await?

File: packages/cli/package.json [hunk 2]
```diff
@@ -80,6 +82,7 @@
     "lodash": "^4.17.15",
     "multiaddr": "^10.0.1",
     "peer-id": "^0.15.3",
+    "prom-client": "^14.0.1",
```
Comment (dapplion) [inline packages/cli/package.json:85]: Why is `"prom-client"` a dependency of the CLI?

File: packages/cli/src/cmds/beacon/handler.ts [hunk 3]
```diff
@@ -58,17 +55,12 @@ export async function beaconHandler(args: IBeaconArgs & IGlobalArgs): Promise<vo
...
-  if (options.metrics.enabled) {
-    dbMetrics = createDbMetrics();
-    metricsRegistries.push(dbMetrics.registry);
-  }
+  const metricsRegistries: Registry[] = [];
   const db = new BeaconDb({
     config,
     controller: new LevelDbController(options.db, {logger: logger.child(options.logger.db)}),
-    metrics: dbMetrics?.metrics,
+    metrics: options.metrics.enabled,
```
Comment (dapplion) [inline packages/cli/src/cmds/beacon/handler.ts:63]: So now `createDbMetrics()` is called in BeaconDb, how are they then wired to the exposed metrics server?

File: packages/cli/src/cli.ts [hunk 4]
```diff
@@ -15,12 +17,14 @@ const bottomBanner = `📖 For more information, check the CLI reference:
...
 
+export const yarg = yargs((hideBin as (args: string[]) => string[])(process.argv));
+
 /**
  * Common factory for running the CLI and running integration tests
  * The CLI must actually be executed in a different script
  */
 export function getLodestarCli(): yargs.Argv {
-  const lodestar = yargs
+  const lodestar = yarg
```
Comment (dapplion) [inline packages/cli/src/cli.ts:27]: So what changed with respect to ESM for yargs? The name `yarg` is here for a lack of a better one right?

File: packages/cli/src/validatorDir/ValidatorDir.ts [hunk 5]
```diff
@@ -71,6 +72,10 @@ export class ValidatorDir {
     }
   }
 
+  static async create(baseDir: string, pubkey: string, options?: IValidatorDirOptions): Promise<ValidatorDir> {
+    return new ValidatorDir(baseDir, pubkey, options);
```
Comment (dapplion) [inline packages/cli/src/validatorDir/ValidatorDir.ts]: What the reason for an async function that instantiates the ValidatorDir class? It turns a bunch of functions async downstream

File: packages/fork-choice/test/perf/protoArray/computeDeltas.test.ts [hunk 6]
```diff
@@ -6,11 +6,11 @@ import {
   EffectiveBalanceIncrements,
   getEffectiveBalanceIncrementsZeroed,
 } from "@chainsafe/lodestar-beacon-state-transition";
-import {generatePerfTestCachedStateAltair} from "@chainsafe/lodestar-beacon-state-transition/test/perf/util";
+import {generatePerfTestCachedStateAltair} from "../../../../beacon-state-transition/test/perf/util.js";
```
Comment (dapplion) [inline packages/fork-choice/test/perf/protoArray/computeDeltas.test.ts:9]: Yeah it looks nasty hahaha now I see why the test-utils package

File: packages/keymanager-server/package.json [hunk 7]
```diff
@@ -23,8 +25,9 @@
     "build:lib:watch": "yarn run build:lib --watch",
     "build:typedocs": "typedoc --exclude src/index.ts --out typedocs src",
     "build:types:watch": "yarn run build:types --watch",
+    "check-build": "node -e \"(async function() { await import('./lib/index.js') })()\"",
     "check-types": "tsc",
-    "lint": "eslint --color --ext .ts src/ test/",
+    "lint": "eslint --color --ext .ts src/",
```
Comment (dapplion) [inline packages/keymanager-server/package.json:30]: Why not lint test anymore?

File: packages/keymanager-server/src/util/lockfile.ts [hunk 8]
```diff
@@ -1,24 +1,20 @@
-type Lockfile = {
+export type Lockfile = {
   lockSync(path: string): void;
   unlockSync(path: string): void;
 };
 
-let lockFile: Lockfile | null = null;
+const lockFile: Lockfile = (await import("lockfile")) as Lockfile;
```
Comment (dapplion) [inline packages/keymanager-server/src/util/lockfile.ts:6]: The point of this file was to import lockfile only when needed. `lockfile` library attaches listeners to the process, so I wanted that to only happen when necessary

File: packages/lodestar/package.json [hunk 9]
```diff
@@ -12,7 +12,25 @@
...
   },
   "version": "0.37.0",
-  "main": "lib/index.js",
+  "type": "module",
+  "exports": {
+    ".": {
+      "import": "./lib/index.js"
+    },
+    "./chain": {
+      "import": "./lib/chain/index.js"
```
Comment (dapplion) [inline packages/lodestar/package.json]: Why is chain exported? Who is using it directly?

File: packages/lodestar/src/db/beacon.ts [hunk 10]
```diff
@@ -47,9 +50,15 @@ export class BeaconDb extends DatabaseService implements IBeaconDb {
...
 
-  constructor(opts: IDatabaseApiOptions) {
-    super(opts);
-    this.metrics = opts.metrics;
+  constructor(opts: Omit<IDatabaseApiOptions, "metrics"> & {metrics?: boolean}) {
+    if (opts.metrics) {
+      const {metrics, registry} = createDbMetrics();
+      super({...opts, metrics});
+      this.metrics = metrics;
+      this.metricsRegistry = registry;
```
Comment (dapplion) [inline packages/lodestar/src/db/beacon.ts:58]: How is this metricsRegistry then connected to the metrics server?

File: packages/lodestar/src/db/repositories/block.ts [hunk 11]
```diff
@@ -1,7 +1,7 @@
 import {IChainForkConfig} from "@chainsafe/lodestar-config";
 import {Bucket, Db, IDbMetrics, Repository} from "@chainsafe/lodestar-db";
-import {allForks, ssz} from "@chainsafe/lodestar-types";
-import {getSignedBlockTypeFromBytes} from "../../util/multifork";
```
Comment (dapplion) [inline packages/lodestar/src/db/repositories/block.ts:4]: `getSignedBlockTypeFromBytes` should be in `util/multifork` not `lodestar-beacon-state-transition`

File: packages/lodestar/src/network/gossip/gossipsub.ts [hunk 12]
```diff
@@ -1,8 +1,10 @@
 /* eslint-disable @typescript-eslint/naming-convention */
 import Libp2p from "libp2p";
-import Gossipsub from "libp2p-gossipsub";
-import {GossipsubMessage, SignaturePolicy, TopicStr} from "libp2p-gossipsub/src/types";
-import {PeerScore, PeerScoreParams} from "libp2p-gossipsub/src/score";
+import GossipsubDefault from "libp2p-gossipsub";
+// TODO remove once Gossipsub goes ESM
+const Gossipsub = ((GossipsubDefault as unknown) as {default: unknown}).default as typeof GossipsubDefault;
```
Comment (dapplion) [inline packages/lodestar/src/network/gossip/gossipsub.ts:5]: Note to open an issue for this TODO

File: packages/lodestar/src/network/gossip/scoringParameters.ts [hunk 13]
```diff
@@ -1,13 +1,17 @@
...
-import {getActiveForks} from "../forks";
-import {Eth2GossipsubModules} from "./gossipsub";
-import {GossipType} from "./interface";
-import {stringifyGossipTopic} from "./topic";
+import {PeerScoreThresholds} from "libp2p-gossipsub/src/score/index.js";
+import {
+  defaultTopicScoreParams,
+  PeerScoreParams,
+  TopicScoreParams,
+} from "libp2p-gossipsub/src/score/peer-score-params.js";
```
Comment (dapplion) [inline packages/lodestar/src/network/gossip/scoringParameters.ts:9]: TEMP imports, change after gossipsub goes ESM

**Comment by dapplion** [comment]
Note, temp imports to change after gossipsub going ESM

https://github.com/ChainSafe/lodestar/blob/5d086b58a20f57c69f1d2cc90d5bf785f851f535/packages/lodestar/src/network/gossip/gossipsub.ts#L4-L5

https://github.com/ChainSafe/lodestar/blob/5d086b58a20f57c69f1d2cc90d5bf785f851f535/packages/lodestar/src/network/gossip/scoringParameters.ts#L4

https://github.com/ChainSafe/lodestar/blob/5d086b58a20f57c69f1d2cc90d5bf785f851f535/packages/lodestar/src/network/gossip/validation/index.ts#L1

And more, search for `"libp2p-gossipsub/src`

File: packages/lodestar/test/e2e/keymanager/keymanager.test.ts [hunk 14]
```diff
@@ -6,17 +6,17 @@ import tmp from "tmp";
 import {createIBeaconConfig, IBeaconConfig, IChainConfig} from "@chainsafe/lodestar-config";
 import {KeymanagerApi, KeymanagerServer} from "@chainsafe/lodestar-keymanager-server";
 import {chainConfig as chainConfigDef} from "@chainsafe/lodestar-config/default";
-import {HttpClient} from "@chainsafe/lodestar-api/src";
-import {getClient} from "@chainsafe/lodestar-api/src/keymanager/client";
+import {HttpClient} from "@chainsafe/lodestar-api";
+import {getClient} from "../../../../api/src/keymanager/client.js";
```
Comment (dapplion) [inline packages/lodestar/test/e2e/keymanager/keymanager.test.ts:10]: Can't use root exports? Why?

**Comment by dapplion** [comment]
Note: Multiple unit tests have been skipped, search for `TODO remove stub` and fix after merging with non-stubby tests

File: packages/params/test/e2e/setPresetOk.ts [hunk 15]
```diff
@@ -2,12 +2,13 @@
 // It demostrates how to properly change the Lodestar preset safely
 
 // 1. Import from @chainsafe/lodestar-params/setPreset only
-import {setActivePreset, PresetName} from "../../setPreset";
+import {setActivePreset, PresetName} from "../../lib/setPreset.js";
```
Comment (dapplion) [inline packages/params/test/e2e/setPresetOk.ts]: Why importing from `lib` and not `src`?

File: packages/types/src/allForks/index.ts [hunk 16]
```diff
@@ -1,2 +1,3 @@
-export * as ts from "./types";
-export * as ssz from "./sszTypes";
+export * from "./types.js";
```
Comment (dapplion) [inline packages/types/src/allForks/index.ts:1]: Why the new line `export * from "./types.js";`?

**Comment by dapplion** [comment]
@wemeetagain One issue I noticed is that if you import from a file without an extension there's no warning anywhere that it's not gonna work. You'll have to wait until running `check-build` to realize it's a bad import. Is there any lint rule or something that ensures correct imports at the IDE level while writing code? I think that's mandatory to ensure we don't go crazy after merging this :pray:

File: packages/beacon-state-transition/.mocharc.yaml (hunk 1 above)
Comment (wemeetagain) [reply packages/beacon-state-transition/.mocharc.yaml:4]: Yeah

File: packages/cli/src/cli.ts (hunk 4 above)
Comment (wemeetagain) [reply packages/cli/src/cli.ts:27]: right, for lack of a better one. It needed a single imported `yargs` instance, broke on multiple instantiations.

File: packages/cli/src/validatorDir/ValidatorDir.ts (hunk 5 above)
Comment (wemeetagain) [reply packages/cli/src/validatorDir/ValidatorDir.ts]: whoops, this should be removed (or kept if you want lazy loading lockfile dependency) (or figure our a different place/way to lazy load lockFile)

File: packages/keymanager-server/src/util/lockfile.ts (hunk 8 above)
Comment (wemeetagain) [reply packages/keymanager-server/src/util/lockfile.ts:6]: With async initialization, it ends up complicating the ValidatorDirectory and builders and such. Thats why I had originally added that `static async create` method.

File: packages/lodestar/package.json (hunk 9 above)
Comment (wemeetagain) [reply packages/lodestar/package.json]: tests

File: packages/lodestar/test/e2e/keymanager/keymanager.test.ts (hunk 14 above)
Comment (wemeetagain) [reply packages/lodestar/test/e2e/keymanager/keymanager.test.ts:10]: Only exports declared in package.json can be used. No more reaching randomly into `src/`.

File: packages/types/src/allForks/index.ts (hunk 16 above)
Comment (wemeetagain) [reply packages/types/src/allForks/index.ts:1]: So we can remove the package-level `allForks.js` and `allForks.d.ts`. This file is now exported at the path `./allForks`.

So `import {BeaconState} from "@chainsafe/lodestar-types/allForks"` imports from this file.

File: packages/cli/package.json (hunk 2 above)
Comment (wemeetagain) [reply packages/cli/package.json:85]: not sure?? will delete

**Comment by wemeetagain** [comment]
> One issue I noticed is that if you import from a file without an extension there's no warning anywhere that it's not gonna work

There should be a linter error already enabled. Our forked https://github.com/ChainSafe/eslint-plugin-node

File: packages/keymanager-server/package.json (hunk 7 above)
Comment (wemeetagain) [reply packages/keymanager-server/package.json:30]: There are no tests, it breaks when it tries to lint a non-existent or empty directory

File: packages/cli/package.json (hunk 2 above)
Comment (wemeetagain) [reply packages/cli/package.json:85]: Oh, its there for the `Registry` type, used in the beacon handler

File: packages/lodestar/src/db/beacon.ts (hunk 10 above)
Comment (wemeetagain) [reply packages/lodestar/src/db/beacon.ts:58]: I think it got removed in a merge from master, adding it back
EDIT: It wasn't ever wired up

File: packages/lodestar/package.json (hunk 9 above)
Comment (dapplion) [reply packages/lodestar/package.json]: So requiring from src doesn't work? It doesnt feel super clean to have an extra export for something that not used outside of tests

Comment (wemeetagain) [reply packages/lodestar/package.json]: relative path imports work. I'll remove this

**Review by g11tech — APPROVED** [review]

**Pull request merged**